File: scripts/tests/test-check-test-schema-dup.js
#!/usr/bin/env node
/**
 * test-check-test-schema-dup.js
 * scripts/check-test-schema-dup.js の planted violation テスト（柱3-e）。
 * MOCK_CHANGED_FILES と MOCK_FILE_*_BASE/_HEAD で git 非依存に検証。
 */
'use strict';
const { spawnSync } = require('child_process');
const path = require('path');
const SCRIPT = path.resolve(__dirname, '..', 'check-test-schema-dup.js');

let passed = 0, failed = 0;
function check(label, cond, detail = '') {
  if (cond) { console.log(`  PASS: ${label}`); passed++; }
  else { console.error(`  FAIL: ${label}${detail ? ' — ' + detail : ''}`); failed++; }
}
function safeKey(p) { return p.replace(/[^A-Za-z0-9]/g, '_').toUpperCase(); }
function run(changed, files) {
  const env = { ...process.env, BASE_SHA: 'mock-base', HEAD_SHA: 'mock-head',
    MOCK_CHANGED_FILES: changed.join('\n') };
  for (const [p, v] of Object.entries(files)) {
    env[`MOCK_FILE_${safeKey(p)}_BASE`] = v.base;
    env[`MOCK_FILE_${safeKey(p)}_HEAD`] = v.head;
  }
  const r = spawnSync('node', [SCRIPT], { env, encoding: 'utf8' });
  return { code: r.status, out: (r.stdout || '') + (r.stderr || '') };
}

const REAL_DDL_BASE = 'async def s(conn):\n    await conn.execute(text("""\n        CREATE TABLE IF NOT EXISTS leads (id INT)\n    """))\n';
const REAL_DDL_ADD = REAL_DDL_BASE + '    await conn.execute(text("""\n        CREATE TABLE IF NOT EXISTS staff (id INT)\n    """))\n';
const TRAP = 'def test_x():\n    sql = "CREATE TABLE foo (id INT)"\n    assert "CREATE TABLE a" in sql\n';

// 1: 本物の複製を新規追加 → 赤
let r1 = run(['backend/tests/test_x.py'],
  { 'backend/tests/test_x.py': { base: REAL_DDL_BASE, head: REAL_DDL_ADD } });
check('充足版: 本物のCREATE TABLE新規追加で赤(exit1)', r1.code === 1, `code=${r1.code}`);

// 2: 罠（sql=/assert）を追加しても複製ではない → 緑
let r2 = run(['backend/tests/test_x.py'],
  { 'backend/tests/test_x.py': { base: REAL_DDL_BASE, head: REAL_DDL_BASE + TRAP } });
check('罠版: sql=/assert追加は複製でない→緑(exit0)', r2.code === 0, `code=${r2.code}`);

// 3: 変更なし（同一）→ 緑
let r3 = run(['backend/tests/test_x.py'],
  { 'backend/tests/test_x.py': { base: REAL_DDL_BASE, head: REAL_DDL_BASE } });
check('不変版: 増減なし→緑(exit0)', r3.code === 0, `code=${r3.code}`);

// 4: conftest.py は対象外 → 緑
let r4 = run(['backend/tests/conftest.py'],
  { 'backend/tests/conftest.py': { base: REAL_DDL_BASE, head: REAL_DDL_ADD } });
check('除外版: conftest.pyは対象外→緑(exit0)', r4.code === 0, `code=${r4.code}`);

// 5: プレースホルダ {schema} の本物追加 → 赤（変種取りこぼさない・柱3-b）
const PH_BASE = 'async def s(conn):\n    await conn.execute(text(f"""\n        CREATE TABLE IF NOT EXISTS {schema}.leads (id INT)\n    """))\n';
const PH_ADD = PH_BASE + '    await conn.execute(text(f"""\n        CREATE TABLE IF NOT EXISTS {schema}.staff (id INT)\n    """))\n';
let r5 = run(['backend/tests/test_ph.py'],
  { 'backend/tests/test_ph.py': { base: PH_BASE, head: PH_ADD } });
check('変種版: プレースホルダ{schema}の複製追加で赤(exit1)', r5.code === 1, `code=${r5.code}`);

console.log('');
console.log(`結果: PASS=${passed} FAIL=${failed}`);
if (failed === 0) console.log('ALL PASS');
process.exit(failed === 0 ? 0 : 1);

File: scripts/tests/test_check_test_schema_dup.py
#!/usr/bin/env python3
"""
柱3-a/b ペアテスト。

合成ケース(欠落版/充足版)と、実データ全件検算の両方を必ず実行する。
自作の練習問題だけで済ませることは禁止(2026-07-24 の取りこぼし再発防止)。
"""
import os
import subprocess
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, os.path.dirname(HERE))
import check_test_schema_dup as m

Q = '"""'
FAILED = []


def check(label, got, want):
    if got == want:
        print("PASS {}: {}".format(label, got))
    else:
        print("FAIL {}: got={} want={}".format(label, got, want))
        FAILED.append(label)


def names(src):
    return [n for _, n in m.tables_in_source(src)]


def assign(body):
    return "_X_DDL = " + Q + "\n" + body + "\n" + Q + "\n"


def synthetic():
    check("form1-assign",
          names(assign("CREATE TABLE leads (id INTEGER)")), ["leads"])

    src2 = ("conn.execute(\n    text(\n        " + Q + "\n"
            "CREATE TABLE staff (id INTEGER)\n" + Q + "\n    )\n)\n")
    check("form2-multiline", names(src2), ["staff"])

    src3 = ("conn.exec_driver_sql(" + Q + "\n"
            "CREATE TABLE audit_logs (id INTEGER)\n" + Q + ")\n")
    check("form3-exec-driver", names(src3), ["audit_logs"])

    src4 = ("dbapi_conn.execute(" + Q + "\n"
            "CREATE TABLE tenants (id INTEGER)\n" + Q + ")\n")
    check("form4-dbapi", names(src4), ["tenants"])

    check("ifne-not-if",
          names(assign("CREATE TABLE IF NOT EXISTS public.suppliers (id INTEGER)")),
          ["public.suppliers"])

    srcf = ("_T = 'leads'\n_S = f" + Q + "\n"
            "CREATE TABLE IF NOT EXISTS {_T} (id INTEGER)\n" + Q + "\n")
    check("fstring-placeholder", names(srcf), ["{expr}"])

    check("as-syntax-excluded",
          names(assign("CREATE TABLE public.tenants AS SELECT 1")), [])

    check("trap-docstring",
          names(Q + "\nCREATE TABLE foo (id INTEGER)\n" + Q + "\n"), [])
    check("trap-assert",
          names("assert 'CREATE TABLE bar' in sql\n"), [])
    check("trap-comment",
          names("# CREATE TABLE baz\nx = 1\n"), [])


def pair_exit():
    base_src = assign("CREATE TABLE leads (id INTEGER)")
    head_src = base_src + assign("CREATE TABLE staff (id INTEGER)")
    orig_changed = m.changed_test_files
    orig_file_at = m.file_at
    os.environ["BASE_SHA"] = "BASE"
    os.environ["HEAD_SHA"] = "HEAD"
    m.changed_test_files = lambda b, h: ["backend/tests/test_pair_probe.py"]
    try:
        m.file_at = lambda sha, path: head_src if sha == "HEAD" else base_src
        check("pair-violation-exit1", m.main(), 1)
        m.file_at = lambda sha, path: base_src
        check("pair-clean-exit0", m.main(), 0)
    finally:
        m.changed_test_files = orig_changed
        m.file_at = orig_file_at
        os.environ.pop("BASE_SHA", None)
        os.environ.pop("HEAD_SHA", None)


def real_data():
    sha = subprocess.run(["git", "rev-parse", "HEAD"],
                         capture_output=True, text=True, check=True).stdout.strip()
    files = m.list_test_files(sha)
    check("real-file-list-nonempty", len(files) > 0, True)
    for trap in sorted(m.EXCLUDE_FILES):
        check("real-excluded " + trap, trap in files, False)
    parse_failed = []
    if_names = []
    hit_files = 0
    total = 0
    for path in sorted(files):
        try:
            found = m.tables_in_source(m.file_at(sha, path))
        except Exception as exc:
            parse_failed.append("{}: {}".format(path, exc))
            continue
        if found:
            hit_files += 1
            total += len(found)
        for _, name in found:
            if name.upper() == "IF":
                if_names.append(path)
    check("real-parse-failed", parse_failed, [])
    check("real-if-zero", if_names, [])
    print("REAL SUMMARY sha={} files={} tables={}".format(sha, hit_files, total))


def main():
    synthetic()
    pair_exit()
    real_data()
    if FAILED:
        print("RESULT: FAIL " + ",".join(FAILED))
        return 1
    print("RESULT: ALL PASS")
    return 0


if __name__ == "__main__":
    sys.exit(main())
